enhancements/sig-architecture/14-addon-cluster-proxy/README.md

```diff
@@ -0,0 +1,381 @@
+                     |                                       |
+  ----------------   | ----------------     ---------------  |
+  | Proxy Client |> > >| Proxy Server | <===| Proxy Addon |  |
+  ----------------   | ----------------     |  Controller |  |
+                     |      V               ---------------  |
+                     |      V                    ‖           |
+                     |    / V \                  ‖           |
+                     ----/| V |\-----------------‖------------
+                          | V |                  ‖
+                   Reverse| V |Tunnel            ‖ Install/Operate
+                          | V |                  ‖ 
+ -------------------------| V |-------------     ‖
+ | Managed-Cluster        | V |            |     ‖
+ |                        | v |            |     ‖
+ | -----------------   ---------------     |     ‖
+ | | KubeApiserver |< <| Proxy Agent | <==========
+ | -----------------   ---------------     |
+ -------------------------------------------
+
+```
+
+- A proxy-agent that establishes a long connection towards the proxy-server
+  and register itself w/ the cluster's identity.
+- A proxy-server that receives handshakes from proxy-agent and maintains 
+  tunnel over the long connection.
+- A proxy-client requesting the target kube-apiserver through the tunnel
+  over either websocket HTTP-CONNECT or grpc protocol.
+- An addon controller automates the installation of proxy servers and proxy 
+  agents by consuming the corresponding configuration custom resource. And
+  optionally the proxy-agents can aggregate and report the status upwards
+  so that admins can checkout their healthiness easily w/o even touching
+  the managed-clusters.
+
+
+#### Components
+
+##### Cluster Proxy Operator
+
+According to the latest design and implementation of [addon-framework](https://github.com/open-cluster-management-io/enhancements/tree/main/enhancements/sig-architecture/8-addon-framework),
+we will bring a new agent controller named `cluster-proxy-operator` that
+manages the lifecycle of the proxy components by watching the configuration
+under a new group of api `proxy.open-cluster-management.io/v1alpha1`:
+
+- __ManagedProxyServer__: the configuration and status of proxy-servers.
+
+- __MangedProxyAgent__: the configuration and status of proxy-agents.
+
+Additionally, the operator will be list-watching the standard `ManagedClusterAddon`
+resource and automatically installs proxy-agents w/ the expected options
+upon discovering the addon in each cluster's namespace.
+
+
+##### ManagedProxyServer
+
+The `ManagedProxyServer` is supposed to be a cluster-scoped resource and holds 
+information for provisioning an active proxy-server instance. An example will 
+be:
+
+```yaml
+apiVersion: proxy.open-cluster-management.io/v1alpha1
+kind: ManagedProxyServer
+metadata:
+  name: default # Or any unique name, will be the name of the workload
+spec:
+  replicas: 1
```

Comment:
i presume multiple replicas will be increasing both high-availability and throughput:

- the proxy-agents is working in simply stateless active-active replicas b/c proxy-servers will be distributing traffic to those registered agents (currently in round-robin strategy).
- the proxy-servers' scaling is a bit tricky. initially proxy-agents doesnt know the actual # of proxy-servers' replicas behind the entry load-balancer, so during the process when proxy-agent is registering itself to the any of the proxy-server, it will be told the identity (a UUID by default) of the proxy-server and also an expected # of replicas via the grpc protocol. so the proxy-agent will be keep attempting the registration until it was successfully registerred to the expected number of servers.

given that all the replicas of proxy-servers and proxy-agents are active, i think scaling up will not only improve the failure/shutdown-tolerance but also the throughput (if the tunnel is working over public internet, the throughput will be converging to a bottleneck due to the bandwidth i think..)